File: AOC-2024/day22/2/main.py
with open('input.txt') as infile:
    numbers = [int(n) for n in infile.read().strip().split('\n')]

def mix(res, num):
    return res ^ num

def prune(res):
    return res % 16777216

def next(num):
    # step 1
    num = prune(mix(num * 64, num))
    # step 2
    num = prune(mix(num // 32, num))
    #step 3
    num = prune(mix(num * 2048, num))
    return num

# n = 123
# print(f'testing with {n=}: {n%10}')
# window = 0
# for i in range(10):
#     nn = next(n)
#     delta = nn % 10 - n % 10
#     window <<= 5
#     window |= abs(delta) | (0b10000 if delta < 0 else 0b00000)
#     window &= 0b1111_1111_1111_1111_1111 # 20 digits = 4 * 5 digits
#     b = f'{window:20b}'
#     print(f'  {i}. n={nn:8}: {nn % 10} ({delta:2}) window: {b[0:5]} {b[5:10]} {b[10:15]} {b[15:20]}')
#     n = nn
# print('-'*20)

print('generating sequences...')
seqs = []
all_wins = set()
for i, num in enumerate(numbers):
    if i % 100 == 0:
        print('.', end='', flush=True)
    n = num
    window = 0
    seq = {}
    for j in range(2000):
        nn = next(n)
        delta = nn % 10 - n % 10
        window <<= 5
        window |= abs(delta) | (0b10000 if delta < 0 else 0b00000)
        window &= 0b1111_1111_1111_1111_1111 # 20 digits = 4 * 5 digits
        n = nn
        if j >= 4: # cant sell before
            if window not in seq: # if it was we would have stopped before and never reached this
                seq[window] = nn % 10
                all_wins.add(window)
    seqs.append(seq)
print()
print(f'detected {len(all_wins)} windows in 2000 iterations over {len(numbers)} inputs')
print('comparing windows...')
max_win = 0
max_profit = 0
for i, window in enumerate(all_wins):
    if i % 1000 == 0:
        if i % 10_000 == 0:
            print('|', end='', flush=True)
        else:
            print('.', end='', flush=True)
    profit = 0
    for seq in seqs:
        if window in seq:
            profit += seq[window]
    if profit > max_profit:
        max_profit = profit
        max_win = window
print('done!')
b = f'{max_win:20b}'
print(f'max_win = {b[0:5]} {b[5:10]} {b[10:15]} {b[15:20]} for {max_profit} bananas')
w0 = ((max_win >> 15) & 0b01111) * (-1 if ((max_win >> 15) & 0b10000) > 0 else 1)
w1 = ((max_win >> 10) & 0b01111) * (-1 if ((max_win >> 10) & 0b10000) > 0 else 1)
w2 = ((max_win >>  5) & 0b01111) * (-1 if ((max_win >>  5) & 0b10000) > 0 else 1)
w3 = ((max_win >>  0) & 0b01111) * (-1 if ((max_win >>  0) & 0b10000) > 0 else 1)
print(f'sequence: {w0:2} {w1:2} {w2:2} {w3:2}')

File: AOC-2024/day22/2/main.c
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>

static inline uint64_t mix(uint64_t result, uint64_t number) {
    return result ^ number;
}

static inline uint64_t prune(uint64_t result) {
    return result % 16777216;
}

static inline uint64_t next(uint64_t number) {
    number = prune(mix(number * 64, number));
    number = prune(mix(number / 32, number));
    number = prune(mix(number * 2048, number));
    return number;
}


int main() {
    // 2<<19 reserves space for all 2**20 possibel values, no compacting needed
    // map window -> total babanas; zeroed out with calloc
    uint64_t* map = calloc(2 << 19, sizeof(uint64_t));
    // set keeping track which window was already encountered this time; zeroed out with calloc
    size_t* mask = calloc(2 << 19, sizeof(size_t));

    uint64_t max_window = 0;
    uint64_t max_profit = 0;

    uint64_t* numbers = malloc(sizeof(uint64_t));
    size_t n_length = 0;
    size_t n_capacity = 1;

    FILE* infile = fopen("input.txt", "r");
    uint64_t number;
    while (fscanf(infile, "%"PRIu64, &number) != EOF) {
        if (n_length == n_capacity) {
            n_capacity *= 2;
            numbers = realloc(numbers, sizeof(uint64_t) * n_capacity);
        }
        numbers[n_length++] = number;
    }

    printf("read %zu input numbers\n", n_length);

    for (size_t i = 0;i < n_length;i++) {
        uint64_t n = numbers[i];
        uint64_t window = 0;
        for (size_t j = 0;j < 2000;j++) {
            uint64_t nn = next(n);
            int64_t delta = nn % 10 - n % 10;
            window <<= 5;
            window |= abs(delta) | (delta < 0 ? 0b10000 : 0b00000);
            window &= 0b11111111111111111111; // 5*4
            n = nn;
            if (j < 4) continue; // window not initially filled yet
            if (mask[window] != i+1) { // this window has not yet appeared for this input
                mask[window] = i+1; // set mask for this window for this input
                map[window] += nn % 10; // add bananas
                if (map[window] > max_profit) {
                    max_profit = map[window];
                    max_window = window;
                }
            }
        }
    }

    printf("maximum possible profit is %"PRIu32" bananas\n", max_profit);
    printf("using window %020b\n", max_window);
}
